File: memex_hermes/envelope.py
"""JSON envelope contract between the Python MemoryProvider and the memex-hermes binary.

Mirrors `src/core/envelope.ts` exactly. Drift between the two sides is a
contract bug — when one moves, the other moves with it (the integration test
under `test/e2e/` round-trips an envelope through both sides to catch drift).

The base shape extends memex-core's `HookInput` (`hook_event_name` plus optional
`session_id`, `cwd`, `transcript_path`). Hermes events carry an `args` payload
narrowed at the receiver by the `hook_event_name` discriminator. Per-event
outputs are JSON-serializable shapes the binary writes back on stdout.

Strict typing notes:
- `args` and `metadata` are `Mapping[str, Any]` at the envelope boundary because
  Hermes can pass arbitrary kwargs/metadata; the receiver narrows them via the
  per-event `TypedDict`s declared below.
- Per CLAUDE.md, `Any` is permitted at the Hermes ABC boundary and is narrowed
  before entering internal code.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any, Final, Literal, TypedDict

from typing_extensions import NotRequired

# ---- Event-name constants (single source of truth) -------------------------

HERMES_HEALTH: Final = "Hermes.health"
HERMES_INIT: Final = "Hermes.init"
HERMES_SYSTEM_PROMPT: Final = "Hermes.system-prompt"
HERMES_PREFETCH: Final = "Hermes.prefetch"
HERMES_QUEUE_PREFETCH: Final = "Hermes.queue-prefetch"
HERMES_SYNC_TURN: Final = "Hermes.sync-turn"
HERMES_SESSION_END: Final = "Hermes.session-end"
HERMES_PRE_COMPRESS: Final = "Hermes.pre-compress"
HERMES_MEMORY_WRITE: Final = "Hermes.memory-write"
HERMES_SESSION_SWITCH: Final = "Hermes.session-switch"
HERMES_SHUTDOWN: Final = "Hermes.shutdown"
HERMES_TOOL_SEARCH: Final = "Hermes.tool-search"
HERMES_TOOL_REMEMBER: Final = "Hermes.tool-remember"
HERMES_TOOL_RECALL: Final = "Hermes.tool-recall"

HermesEventName = Literal[
    "Hermes.health",
    "Hermes.init",
    "Hermes.system-prompt",
    "Hermes.prefetch",
    "Hermes.queue-prefetch",
    "Hermes.sync-turn",
    "Hermes.session-end",
    "Hermes.pre-compress",
    "Hermes.memory-write",
    "Hermes.session-switch",
    "Hermes.shutdown",
    "Hermes.tool-search",
    "Hermes.tool-remember",
    "Hermes.tool-recall",
]

HermesAgentContext = Literal["primary", "subagent", "cron", "flush"]
HermesMemoryAction = Literal["add", "replace", "remove"]
HermesMemoryTarget = Literal["memory", "user"]
HermesToolScope = Literal["session", "project", "global"]

# ---- Per-event argument shapes --------------------------------------------


class HermesInitArgs(TypedDict, total=False):
    """`Hermes.init` args — initialize kwargs forwarded from the framework.

    `hermes_home`, `platform`, `agent_context` are always present (auto-injected
    by the framework + agent_init.py); the remaining keys are platform / gateway
    specific. Forward-compat: unknown keys are tolerated.
    """

    hermes_home: str
    platform: str
    agent_context: HermesAgentContext
    agent_identity: str
    agent_workspace: str
    parent_session_id: str
    user_id: str
    user_name: str
    session_title: str
    chat_id: str
    chat_name: str
    chat_type: str
    thread_id: str
    gateway_session_key: str


class HermesPrefetchArgs(TypedDict):
    query: str
    session_id: NotRequired[str]


class HermesQueuePrefetchArgs(TypedDict):
    query: str
    session_id: NotRequired[str]


class HermesSyncTurnArgs(TypedDict):
    user_content: str
    assistant_content: str
    session_id: NotRequired[str]


class HermesSessionEndArgs(TypedDict):
    messages: Sequence[Mapping[str, Any]]


class HermesPreCompressArgs(TypedDict):
    messages: Sequence[Mapping[str, Any]]


class HermesMemoryWriteArgs(TypedDict):
    action: HermesMemoryAction
    target: HermesMemoryTarget
    content: str
    metadata: NotRequired[Mapping[str, Any]]


class HermesSessionSwitchArgs(TypedDict):
    new_session_id: str
    parent_session_id: NotRequired[str]
    reset: NotRequired[bool]


class HermesToolSearchArgs(TypedDict):
    query: str
    limit: NotRequired[int]
    types: NotRequired[Sequence[str]]


class HermesToolRememberArgs(TypedDict):
    content: str
    scope: NotRequired[HermesToolScope]
    projectName: NotRequired[str]


class HermesToolRecallArgs(TypedDict):
    name: NotRequired[str]
    limit: NotRequired[int]


# ---- Input envelope -------------------------------------------------------


class HermesHookInput(TypedDict, total=False):
    """The JSON envelope written to the binary's stdin.

    `hook_event_name` is required; `args` is per-event (narrowed by receiver).
    """

    hook_event_name: HermesEventName
    session_id: str
    cwd: str
    transcript_path: str
    args: Mapping[str, Any]


# ---- Per-event output shapes ---------------------------------------------


class HermesHealthOutput(TypedDict):
    ready: bool
    reason: NotRequired[str]


class HermesInitOutput(TypedDict):
    ok: Literal[True]


class HermesSystemPromptOutput(TypedDict):
    block: str


class HermesPrefetchOutput(TypedDict, total=False):
    additionalContext: str


class HermesEmptyOutput(TypedDict):
    """`{}` — used by fire-and-forget events that have no payload to return."""


class HermesSyncTurnOutput(TypedDict, total=False):
    ok: Literal[True]
    mirrored: Sequence[HermesMemoryTarget]


class HermesSessionEndOutput(TypedDict):
    written: int


class HermesPreCompressOutput(TypedDict, total=False):
    summary: str


class HermesMemoryWriteOutput(TypedDict, total=False):
    committed: bool
    suppressed: str


class HermesSessionSwitchOutput(TypedDict):
    ok: Literal[True]


class HermesShutdownOutput(TypedDict):
    ok: Literal[True]


class HermesToolSearchResult(TypedDict, total=False):
    name: str
    type: str
    score: float
    location: str
    snippet: str


class HermesToolSearchOutput(TypedDict):
    results: Sequence[HermesToolSearchResult]


class HermesToolRememberOutput(TypedDict):
    written: str
    synced: bool


class HermesToolRecallEntry(TypedDict):
    name: str
    content: str


class HermesToolRecallOutput(TypedDict):
    entries: Sequence[HermesToolRecallEntry]

File: src/core/envelope.ts
// JSON envelope contract between the Python MemoryProvider and the memex-hermes
// binary. Single source of truth — the Python side (memex_hermes/envelope.py)
// mirrors these shapes as TypedDicts. Drift between the two is a contract bug.
//
// The base HookInput from memex-core carries hook_event_name + a few optional
// top-level fields (session_id, cwd, transcript_path, prompt, tool_name,
// tool_input). Hermes events extend it with an `args` payload narrowed by the
// hook_event_name discriminator. Outputs are JSON-serializable and per-event.

import type { HookInput, HookOutput } from "@jim80net/memex-core";

// ---- Event-name constants (the single source of truth) ----------------------

export const HERMES_EVENTS = {
  HEALTH: "Hermes.health",
  INIT: "Hermes.init",
  SYSTEM_PROMPT: "Hermes.system-prompt",
  PREFETCH: "Hermes.prefetch",
  QUEUE_PREFETCH: "Hermes.queue-prefetch",
  SYNC_TURN: "Hermes.sync-turn",
  SESSION_END: "Hermes.session-end",
  PRE_COMPRESS: "Hermes.pre-compress",
  MEMORY_WRITE: "Hermes.memory-write",
  SESSION_SWITCH: "Hermes.session-switch",
  SHUTDOWN: "Hermes.shutdown",
  TOOL_SEARCH: "Hermes.tool-search",
  TOOL_REMEMBER: "Hermes.tool-remember",
  TOOL_RECALL: "Hermes.tool-recall",
} as const;

export type HermesEventName = (typeof HERMES_EVENTS)[keyof typeof HERMES_EVENTS];

// ---- Per-event argument shapes ---------------------------------------------

export type HermesAgentContext = "primary" | "subagent" | "cron" | "flush";

// initialize kwargs the framework auto-injects (hermes_home/platform/agent_context
// always; the rest may be present depending on the platform/gateway path).
export interface HermesInitArgs {
  hermes_home: string;
  platform: string;
  agent_context: HermesAgentContext;
  agent_identity?: string;
  agent_workspace?: string;
  parent_session_id?: string;
  user_id?: string;
  user_name?: string;
  session_title?: string;
  chat_id?: string;
  chat_name?: string;
  chat_type?: string;
  thread_id?: string;
  gateway_session_key?: string;
  // Forward-compatible: any unknown init kwargs are tolerated and forwarded.
  [extra: string]: unknown;
}

export interface HermesPrefetchArgs {
  query: string;
  session_id?: string;
}

export interface HermesQueuePrefetchArgs {
  query: string;
  session_id?: string;
}

export interface HermesSyncTurnArgs {
  user_content: string;
  assistant_content: string;
  session_id?: string;
}

export interface HermesSessionEndArgs {
  messages: ReadonlyArray<Readonly<Record<string, unknown>>>;
}

export interface HermesPreCompressArgs {
  messages: ReadonlyArray<Readonly<Record<string, unknown>>>;
}

export type HermesMemoryAction = "add" | "replace" | "remove";
export type HermesMemoryTarget = "memory" | "user";

export interface HermesMemoryWriteArgs {
  action: HermesMemoryAction;
  target: HermesMemoryTarget;
  content: string;
  // Provenance dict from Hermes. Common keys: write_origin, execution_context,
  // session_id, parent_session_id, platform, tool_name. Unknown keys tolerated.
  metadata?: Readonly<Record<string, unknown>>;
}

export interface HermesSessionSwitchArgs {
  new_session_id: string;
  parent_session_id?: string;
  reset?: boolean;
}

export interface HermesToolSearchArgs {
  query: string;
  limit?: number;
  types?: ReadonlyArray<string>;
}

export type HermesToolScope = "session" | "project" | "global";

export interface HermesToolRememberArgs {
  content: string;
  scope?: HermesToolScope;
  projectName?: string;
}

export interface HermesToolRecallArgs {
  name?: string;
  limit?: number;
}

// ---- Input envelope ---------------------------------------------------------

// Extends memex-core's HookInput with an `args` payload narrowed by event name.
export interface HermesHookInput<E extends HermesEventName = HermesEventName, A = unknown>
  extends HookInput {
  hook_event_name: E;
  args?: A;
}

// ---- Per-event output shapes -----------------------------------------------

export interface HermesHealthOutput {
  ready: boolean;
  reason?: string;
}

export interface HermesInitOutput {
  ok: true;
}

export interface HermesSystemPromptOutput {
  block: string;
}

// Prefetch reuses memex-core's HookOutput shape ({ additionalContext?: string }).
export type HermesPrefetchOutput = HookOutput;

// Empty acknowledgement for fire-and-forget events.
export type HermesEmptyOutput = Record<string, never>;

export interface HermesSyncTurnOutput {
  ok: true;
  mirrored?: ReadonlyArray<HermesMemoryTarget>;
}

export interface HermesSessionEndOutput {
  written: number;
}

export interface HermesPreCompressOutput {
  summary?: string;
}

export interface HermesMemoryWriteOutput {
  committed: boolean;
  // When the write was dropped (e.g., non-primary agent_context or _session/*
  // project), this names the reason. Implementations log it; spec scenarios
  // assert it via `informational log line` rather than the response field.
  suppressed?: string;
}

export interface HermesSessionSwitchOutput {
  ok: true;
}

export interface HermesShutdownOutput {
  ok: true;
}

export interface HermesToolSearchResult {
  name: string;
  type: string;
  score: number;
  location: string;
  snippet?: string;
}

export interface HermesToolSearchOutput {
  results: ReadonlyArray<HermesToolSearchResult>;
}

export interface HermesToolRememberOutput {
  written: string;
  synced: boolean;
}

export interface HermesToolRecallEntry {
  name: string;
  content: string;
}

export interface HermesToolRecallOutput {
  entries: ReadonlyArray<HermesToolRecallEntry>;
}

// ---- Per-event input/output pairing (compile-time map) ---------------------

export interface HermesEventMap {
  [HERMES_EVENTS.HEALTH]: { args: Record<string, never>; output: HermesHealthOutput };
  [HERMES_EVENTS.INIT]: { args: HermesInitArgs; output: HermesInitOutput };
  [HERMES_EVENTS.SYSTEM_PROMPT]: { args: Record<string, never>; output: HermesSystemPromptOutput };
  [HERMES_EVENTS.PREFETCH]: { args: HermesPrefetchArgs; output: HermesPrefetchOutput };
  [HERMES_EVENTS.QUEUE_PREFETCH]: { args: HermesQueuePrefetchArgs; output: HermesEmptyOutput };
  [HERMES_EVENTS.SYNC_TURN]: { args: HermesSyncTurnArgs; output: HermesSyncTurnOutput };
  [HERMES_EVENTS.SESSION_END]: { args: HermesSessionEndArgs; output: HermesSessionEndOutput };
  [HERMES_EVENTS.PRE_COMPRESS]: { args: HermesPreCompressArgs; output: HermesPreCompressOutput };
  [HERMES_EVENTS.MEMORY_WRITE]: { args: HermesMemoryWriteArgs; output: HermesMemoryWriteOutput };
  [HERMES_EVENTS.SESSION_SWITCH]: {
    args: HermesSessionSwitchArgs;
    output: HermesSessionSwitchOutput;
  };
  [HERMES_EVENTS.SHUTDOWN]: { args: Record<string, never>; output: HermesShutdownOutput };
  [HERMES_EVENTS.TOOL_SEARCH]: { args: HermesToolSearchArgs; output: HermesToolSearchOutput };
  [HERMES_EVENTS.TOOL_REMEMBER]: { args: HermesToolRememberArgs; output: HermesToolRememberOutput };
  [HERMES_EVENTS.TOOL_RECALL]: { args: HermesToolRecallArgs; output: HermesToolRecallOutput };
}
